File: examples/python/speak.py
from __future__ import annotations

import os
import sys
from pathlib import Path

from sipp import (
    ContextRuntimeConfig,
    Endpoint,
    ModelPlacementConfig,
    NativeRuntimeConfig,
    SippClient,
    set_llama_log_quiet,
)

from _support import gpu_layers, int_env


def main() -> None:
    if len(sys.argv) < 4:
        raise SystemExit(
            "usage: python examples/python/speak.py "
            "<model.gguf> <projector.gguf> <output.wav> [text]"
        )
    model_path, projector_path, output_path, *words = sys.argv[1:]
    set_llama_log_quiet(True)

    client = SippClient()
    model = client.models.add([model_path, projector_path])
    client.add("tts", Endpoint.local(model, runtime=runtime_config(4096)))
    speaker_path = os.getenv("SIPP_SPEAKER_AUDIO")
    response = client.speak(
        " ".join(words) or "Hello from Sipp.",
        language=os.getenv("SIPP_LANGUAGE"),
        speaker_audio=Path(speaker_path).read_bytes() if speaker_path else None,
        max_duration_ms=int_env("SIPP_MAX_DURATION_MS"),
    ).result()
    Path(output_path).write_bytes(response["audio"])
    print(
        f'wrote {response["duration_ms"]} ms at '
        f'{response["sample_rate_hz"]} Hz to {output_path}'
    )


def runtime_config(context_size: int) -> NativeRuntimeConfig:
    return NativeRuntimeConfig(
        placement=ModelPlacementConfig(gpu_layers=gpu_layers()),
        context=ContextRuntimeConfig(
            n_ctx=int_env("SIPP_CONTEXT", context_size),
            n_threads=int_env("SIPP_THREADS"),
            n_threads_batch=int_env("SIPP_THREADS"),
            warmup=False,
        ),
    )


if __name__ == "__main__":
    main()

File: examples/node/speak.mjs
import { readFileSync, writeFileSync } from 'node:fs';

import native from '../../lib/node/router.js';
import { gpuLayers, intEnv } from './_support.mjs';

const [modelPath, projectorPath, outputPath, ...words] = process.argv.slice(2);
if (!modelPath || !projectorPath || !outputPath) {
  throw new Error(
    'usage: node examples/node/speak.mjs <model.gguf> <projector.gguf> <output.wav> [text]'
  );
}

const { Endpoint, SippClient, setLlamaLogQuiet } = native;
setLlamaLogQuiet(true);
const client = new SippClient();
const model = await client.models.add([modelPath, projectorPath]);
await client.add('tts', Endpoint.local(model, { runtime: runtimeConfig(4096) }));

const response = await client.speak({
  text: words.join(' ') || 'Hello from Sipp.',
  language: process.env.SIPP_LANGUAGE,
  speakerAudio: process.env.SIPP_SPEAKER_AUDIO
    ? readFileSync(process.env.SIPP_SPEAKER_AUDIO)
    : undefined,
  maxDurationMs: intEnv('SIPP_MAX_DURATION_MS'),
}).response;
writeFileSync(outputPath, response.audio);
console.log(`wrote ${response.durationMs} ms at ${response.sampleRateHz} Hz to ${outputPath}`);

function runtimeConfig(contextSize) {
  return {
    placement: { gpu_layers: gpuLayers() },
    context: {
      n_ctx: intEnv('SIPP_CONTEXT', contextSize),
      n_threads: intEnv('SIPP_THREADS'),
      n_threads_batch: intEnv('SIPP_THREADS'),
      warmup: false,
    },
  };
}
